### PR TITLE
BUG: Fixed deepcopy of object arrays.

### PR DESCRIPTION
**Pull request by pelson**
Fixes a bug with deepcopy of object arrays in F-order:

```
In [1]: import numpy as np

In [2]: np.__version__       
Out[2]: '1.11.0.dev0+9cc55dc'

In [3]: a = {'a': 1}

In [4]: b = {'b': 2}

In [5]: np.array([[a, b], [a, b]], order='F')
Out[5]: 
array([[{'a': 1}, {'b': 2}],
       [{'a': 1}, {'b': 2}]], dtype=object)

In [6]: import copy

In [7]: copy.deepcopy(np.array([[a, b], [a, b]], order='F'))
Out[7]: 
array([[{'a': 1}, {'a': 1}],
       [{'b': 2}, {'b': 2}]], dtype=object)
```

This was discovered as part of a bug found with [biggus](https://github.com/SciTools/biggus) in https://github.com/SciTools/biggus/issues/157.

### CODE REVIEW

File: numpy/core/tests/test_regression.py [hunk 1]
```diff
@@ -457,7 +457,7 @@ def test_swap_real(self, level=rlevel):
 
     def test_object_array_from_list(self, level=rlevel):
         # Ticket #270
-        np.array([1, 'A', None])  # Should succeed
```
Comment (pelson) [inline numpy/core/tests/test_regression.py]: It does succeed - I just wanted it to make an assertion.
Happy to revert though.

File: numpy/core/src/multiarray/methods.c [hunk 2]
```diff
@@ -1426,21 +1429,44 @@ array_deepcopy(PyArrayObject *self, PyObject *args)
         if (deepcopy == NULL) {
             return NULL;
         }
-        it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)self);
```
Comment (pelson) [inline numpy/core/src/multiarray/methods.c]: The problem is that IterNew is iterating in C order, and optr is therefore getting out of synch.

**Comment by jaimefrio** [comment]
Not that I'm against using the new iterator, but wouldn't you get the same result if you changed this line:

```
_deepcopy_call(it->dataptr, optr, PyArray_DESCR(self), deepcopy, visit);
```

with

```
_deepcopy_call(it->dataptr, it->dataptr, PyArray_DESCR(self), deepcopy, visit);
```

and got rid of all the `optr` code, let the iterator handle all the iteration?

**Comment by pelson** [comment]
> and got rid of all the optr code, let the iterator handle all the iteration?

I think you may be right. I was concerned about the fact that ``it`` is iterating over self, not the result array, but I think that is fine in this case. To be honest, I took this as an opportunity to understand the new iterator API - do you have a preference on my next steps - obviously, if I take the easier solution it will be easier to apply to a potential v1.10.1, but on the other hand if I stick with my original implementation it is making use of the new iterator API...

File: numpy/core/src/multiarray/methods.c [hunk 3]
```diff
@@ -1426,21 +1429,44 @@ array_deepcopy(PyArrayObject *self, PyObject *args)
...
+        if (iter == NULL) {
             Py_DECREF(deepcopy);
             return NULL;
         }
-        optr = PyArray_DATA(ret);
-        while(it->index < it->size) {
-            _deepcopy_call(it->dataptr, optr, PyArray_DESCR(self), deepcopy, visit);
-            optr += PyArray_DESCR(self)->elsize;
-            PyArray_ITER_NEXT(it);
+        NpyIter_IterNextFunc *iternext = NpyIter_GetIterNext(iter, NULL);
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/methods.c]: This has to be declared at the beginning of the function to comply with the C89 strictness of MSVC.

**Comment by jaimefrio** [comment]
My personal view is that, for the new iterator to replace the old one, it needs a middle layer that makes implementing simple iterations like this one much simpler and less verbose: the alternative to the 26 lines that your PR adds to `methods.c` is to keep using the old iterator and remove 3.  And there is no obvious benefit from the change, so I would not have included the new iterator as part of this PR.

Then again, this is your PR, not mine, and there is nothing fundamentally wrong with using `NpyIter`. It certainly should be encouraged for new development.  So if you manage to fix the test errors, I wouldn't mind merging it as is.

File: numpy/core/tests/test_regression.py [hunk 4]
```diff
@@ -2080,6 +2080,22 @@ def test_deepcopy_on_0d_array(self):
...
+        # Ticket #6456.
+        a = {'a': 1}
+        b = {'b': 2}
+        arr = np.array([[a, b], [a, b]], order='F')
+        arr_cp = copy.deepcopy(arr)
+
+        assert_equal(arr, arr_cp)
+        self.assertTrue(arr is not arr_cp)
+        # Ensure that we have actually copied the item.
+        self.assertIsNot(arr[0, 1], arr_cp[1, 1])
```
Comment (charris) [inline numpy/core/tests/test_regression.py]: `assertIs` and `assertIsNot` are not available in Python 2.6, use `assertTrue` and `assertFalse` with proper arguments.

File: numpy/core/src/multiarray/methods.c (hunk 3 above)
Comment (charris) [inline numpy/core/src/multiarray/methods.c]: All declarations must go at the beginning of a block.

**Comment by pelson** [comment]
Thanks @jaimefrio and @charris - I've updated the PR based on your review. I've kept the new API implementation, but am still not averse to choosing the easier solution of updating the existing implementation (either way, we can keep the new tests).

File: numpy/core/src/multiarray/methods.c [hunk 5]
```diff
@@ -1406,16 +1406,23 @@ _deepcopy_call(char *iptr, char *optr, PyArray_Descr *dtype,
...
+    npy_intp stride, count;
+
     PyObject *copy, *deepcopy;
-    PyArrayObject *ret;
 
     if (!PyArg_ParseTuple(args, "O", &visit)) {
         return NULL;
     }
-    ret = (PyArrayObject *)PyArray_NewCopy(self, NPY_KEEPORDER);
+    copied_array = (PyArrayObject*) PyArray_NewCopy(self, NPY_KEEPORDER);
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/methods.c]: We should be checking that `copied_array` is not `NULL`, and if it is, return `NULL` to signal an error happened.

File: numpy/core/src/multiarray/methods.c [hunk 6]
```diff
@@ -1426,21 +1433,44 @@ array_deepcopy(PyArrayObject *self, PyObject *args)
         if (deepcopy == NULL) {
             return NULL;
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/methods.c]: Need to `Py_DECREF(copied_array)` before returning.

File: numpy/core/src/multiarray/methods.c [hunk 7]
```diff
@@ -1426,21 +1433,44 @@ array_deepcopy(PyArrayObject *self, PyObject *args)
...
+        iter = (NpyIter *)NpyIter_New(copied_array,
+                                      (NPY_ITER_READWRITE |
+                                       NPY_ITER_EXTERNAL_LOOP |
+                                       NPY_ITER_REFS_OK),
+                                      NPY_KEEPORDER,
+                                      NPY_NO_CASTING,
+                                      NULL);
+        if (iter == NULL) {
             Py_DECREF(deepcopy);
             return NULL;
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/methods.c]: Need to `Py_DECREF(copied_array)` before returning.

File: numpy/core/src/multiarray/methods.c [hunk 8]
```diff
@@ -1406,16 +1406,23 @@ _deepcopy_call(char *iptr, char *optr, PyArray_Descr *dtype,
...
-    PyArrayObject *ret;
 
     if (!PyArg_ParseTuple(args, "O", &visit)) {
         return NULL;
     }
-    ret = (PyArrayObject *)PyArray_NewCopy(self, NPY_KEEPORDER);
+    copied_array = (PyArrayObject*) PyArray_NewCopy(self, NPY_KEEPORDER);
     if (PyDataType_REFCHK(PyArray_DESCR(self))) {
         copy = PyImport_ImportModule("copy");
         if (copy == NULL) {
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/methods.c]: Need to `Py_DECREF(copied_array)` before returning.

File: numpy/core/src/multiarray/methods.c [hunk 9]
```diff
@@ -1426,21 +1433,44 @@ array_deepcopy(PyArrayObject *self, PyObject *args)
...
-        optr = PyArray_DATA(ret);
-        while(it->index < it->size) {
-            _deepcopy_call(it->dataptr, optr, PyArray_DESCR(self), deepcopy, visit);
-            optr += PyArray_DESCR(self)->elsize;
-            PyArray_ITER_NEXT(it);
+        iternext = NpyIter_GetIterNext(iter, NULL);
+        if (iternext == NULL) {
+            NpyIter_Deallocate(iter);
+            Py_DECREF(deepcopy);
+            return NULL;
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/methods.c]: Need to `Py_DECREF(copied_array)` before returning.

**Comment by jaimefrio** [comment]
There seemed to be a potential memory leak in case of failure. Was already there, but would be nice if you couldfix it also as part of this PR.

**Comment by pelson** [comment]
> There seemed to be a potential memory leak in case of failure. Was already there, but would be nice if you could fix it also as part of this PR.

Yep, that seems reasonable! I've added a commit (d158c59). Happy to squash if that is necessary.

**Comment by pelson** [comment]
Additionally, if there are stylistic improvements to the code, do let me know - I'm happy to fix them, and would appreciate critique.

**Comment by pelson** [comment]
Has this missed the v1.11 boat @charris?

**Comment by charris** [comment]
@pelson I'm going to leave this with @jaimefrio for when he gets back. Don't worry, we are shooting for branching 1.12 in April, not that far off.

**Comment by charris** [comment]
Hmm, we should get this in. @pelson, could you rebase this?

**Comment by charris** [comment]
Rebased in #8125, so closing this. Thanks @pelson .

**Comment by pelson** [comment]
Thanks for taking this forward @charris. Great to have it fixed in master 👍

**Comment by aojeda** [comment]
Thanks guys for fixing the issue so quickly! 
Onward!

**Comment by charris** [comment]
@aojeda Thanks, but this is the PR that caused your problem 8-)

**Comment by aojeda** [comment]
Ah ah true, I'll take it back for now then ;-)

**Pull request closed without merging**